### ISSUE TITLE
ELU instead of ReLU in conv_dw_no_bn

### ISSUE THREAD

**Issue by fy-meng** [Daniil-Osokin/lightweight-human-pose-estimation.pytorch#27]
Hi, nice work on this repo! 

I'm wondering that why do you use `ELU` instead of `ReLU` in `con_dw_no_bw`, while the `conv_dw` counterpart uses the regular `ReLU`:
https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch/blob/a6e41cf56e0b5e2d23686de9ef15671833bdb72e/modules/conv.py#L25-L32
Is there a particular reason to use `ELU`? I didn't see any mentioning of activation function in your paper or the original openpose paper.

Thank you!

**Comment by Daniil-Osokin** [comment]
Good question! For this particular network architecture there is no difference, it is from unfinished experiment. Nets with `ReLU` activations exposed "dead" neurons problem, so some percentage of neurons just never works. And the purpose of this work is to do lightweight net, yet maintain baseline accuracy. One of possible ways to reduce network complexity and save original capacity is to get rid of "dead" neurons with other activation function, so network can be more narrow (no "dead" neurons) but have same capacity (and accuracy). So here we've used `ELU` and the next thing is to reduce number of channels in these layers, but the time is left. I've taken this idea from RMNet paper: ["Fast and Accurate Person Re-Identification with RMNet"](https://arxiv.org/pdf/1812.02465.pdf) by E. Izutov.

**Comment by fy-meng** [comment]
Thank you for your explanation!

**Comment by rodyt** [comment]
@Daniil-Osokin You mentioned how `ELU` helps prevent `ReLU` neurons from dying off. I am encountering an issue possibly related to this, and I was wondering if you had run into this problem before, and if so, how you went about fixing it:


I have a built a CNN that takes in an input image and outputs joint confidence heatmaps. In order to verify that my model works correctly, I first try to overfit **one** training batch of my training dataset. 
The logic behind this is that the model (with its millions of weights) should easily be able to overfit the training sample (i.e. I first test the model on an extremely simple and trivial task to see if it works).

I trained the network on this same batch for ~15,000 iterations, using the Adam optimizer and standard hyperparameters.

The issue I am encountering is this: The heatmap is surprisingly **unable** to overfit the _one training example_! (See below)

![heatmapdiagram](https://user-images.githubusercontent.com/36773880/75411846-845a2500-58d5-11ea-8b4a-ce2c7baffcd6.png)

It appears that my model doesn't learn _at all_ and is outputting gibberish!


Next, I tried swapping out the activation function from `ReLU` to `LeakyReLU` in light of your comment. 

After doing so, it appears that the model is at least able to (generally) guess where the joints are, but it is not confident at all in its predictions! (i.e. the activations are very sparse, instead of a dense gaussian kernel like you would expect)

(And even if switching to `LeakyReLU` does help the model learn, I notice that most SOTA models for human pose estimation get excellent performance using just plain `ReLU`. So no matter what, the model should at least be able to overfit _one batch_ with vanilla `ReLU`)


I have verified that my model is connected correctly, and that my input images and ground truth heatmaps are preprocessed correctly.


Any points or suggestions would be greatly appreciated!

**Comment by Daniil-Osokin** [comment]
Hi! It's hard to say without the code. Check if any augmentations involved in dataset and test image included into training batch. Usually (if network has enough layers), switching from `ReLU` to `LeakyReLU` do not have dramatic effect for accuracy. You can also check if loss decreasing, if so, may be testing/visualization code buggy.

**Comment by rodyt** [comment]
@Daniil-Osokin Thanks for the suggestions on what to try!

In literature, I noticed some pose estimation methods use some sort of "_`mask`_" in their loss. After doing some research, I believe this to actually be the source of my error. I would really appreciate it if you could help me understand what this "_`mask`_" is doing and what role it plays in training.

So in a heatmap, a lot of the heatmap values are zero (no joint located at that pixel). The error likely comes from my model learning that the quickest way to reduce the `MSE` is to just output all zeros. 


My understanding is that this "_`mask`_" is applied to the output heatmap of the model that determines which pixels are weighed in the loss. (Quoting from the Openpose paper: "_**W** is a binary mask with **W(**p**)** = 0 when the annotation is missing at the pixel **p**_")


For example, here's some implementations I found on Github:

**mks0601/TF-SimpleHumanPose:**

```
valid_mask = tf.reshape(valid, [cfg.batch_size, 1, 1, cfg.num_kps])
loss = tf.reduce_mean(tf.square(heatmap_outs - gt_heatmap) * valid_mask)
```


**Daniil-Osokin/lightweight-human-pose-estimation.pytorch (this repo):**

```
feedback
def l2_loss(input, target, mask, batch_size):
    loss = (input - target) * mask
    loss = (loss * loss) / 2 / batch_size

    return loss.sum()
```


I also noticed a "_`crowd mask`_" parameter when training on COCO dataset. I'm not sure if this is the same thing as the aforementioned "_`mask`_".

From this repo:

```
for annotation in data['annotations']:
    if annotation['num_keypoints'] != 0 and not annotation['iscrowd']:
```

My question is: What is this "_`mask`_" parameter and how do I implement / use it?

Thanks again for your help

**Comment by Daniil-Osokin** [comment]
COCO has instance segmentation annotations. Some image regions contain tight crowd of persons, so it is not possible to properly label keypoints for each person in a crowd (and COCO annotations just do not have keypoints for them). Such regions have segmentation annotation and [`iscrowd`](http://cocodataset.org/#format-data) attribute set to 1. However network may predict keypoints for that regions (and some of predictions will be valid, because it has persons), but there are no keypoints in annotation. So to not penalize network for predictions in ambiguous regions, these regions are masked out (mask has zeros in place of these regions segmentations), and pixels inside these regions does not contribute to loss.

Despite that, COCO has a lot of images with `iscrowd` == 0 for each persons. So you can overfit to one of them and do not bother yourself with masks handling.

**Comment by rodyt** [comment]
This makes much more sense now.

Do you happen to know what's the performance of the model (accuracy) with and without crowd masking?


By the way, is this crowd mask parameter that you used the same as the aforementioned binary mask mentioned in the OpenPose paper? (In the Openpose paper: "_**W** is a binary mask with **W**(**p**) = 0 when the annotation is missing at the pixel **p**_")

Thanks

**Comment by Daniil-Osokin** [comment]
We have not trained without masking. Regarding the second question, yes I think so, the purpose is the same.

**Comment by rodyt** [comment]
@Daniil-Osokin 

Thanks again for helping me understand. :grinning: 

As an aside, have you experimented with training the model on occluded (i.e. visibility == 1) joints in COCO? If so, what's the performance gain / loss of training with over without (without meaning that you don't try to predict occluded joints at all)?

**Comment by Daniil-Osokin** [comment]
We do not. I'm think the best way is to mask out occluded points too, so network may predict them (if it smart enough), but we will not penalize it if it cannot do it.
